Blocked format api

oicirtap opened the pull request:
Added support for creating blocked Format objects.

* Changed the Format API to use curly-brackets to express blocks
* Changed the Format API to use packBounds vector to express packs
* Updated all relevant tests and files
* Pending: deciding what block metadata to collect and expose during Format instantiation.

stephenchouca commented on include/taco/format.h:
```diff
@@ -65,10 +93,35 @@ class Format {
+  bool isBlocked();
+
+  /// Gets the number of nested block levels in the format
+  int numberOfBlocks();
```
Maybe rename to something like `numBlockLevels`?  The fact that this is referring to the number of levels doesn't really get across in the current name.

stephenchouca commented on include/taco/format.h:
```diff
@@ -145,6 +204,9 @@ class ModeFormat {
 private:
   std::shared_ptr<const ModeFormatImpl> impl;
 
+  bool sizeFixed = false;
```
Should probably go into `ModeFormatImpl`.

stephenchouca commented on include/taco/format.h:
```diff
@@ -145,6 +204,9 @@ class ModeFormat {
   std::shared_ptr<const ModeFormatImpl> impl;
 
+  bool sizeFixed = false;
+  int blockSize;
```
Should probably go into `ModeFormatImpl`.

stephenchouca commented on src/format.cpp:
```diff
@@ -18,39 +18,158 @@ using namespace std;
+        blockSizes[block_i][dim_i] = format.size();
+      } else {
+        taco_uassert(freeSizeBlock[dim_i] == -1) <<
+            "Each dimmension requires at most one free-size block.";
```
"dimension" not "dimmension"

oicirtap commented:
Moved block format metadata to mode_format_impl, and added a SIZE_FIXED format property.
I am not sure if technically singleton mode formats are restricted to always have fixed size 1?

stephenchouca commented on include/taco/lower/mode_format_impl.h:
```diff
@@ -66,13 +66,13 @@ class ModeFormatImpl {
                  bool isBranchless, bool isCompact, bool hasCoordValIter, 
                  bool hasCoordPosIter, bool hasLocate, bool hasInsert, 
-                 bool hasAppend);
+                 bool hasAppend, bool hasFixedSize = false, int size = 0);
```
Maybe the constructor should just take a single argument of type `Dimension` (as defined in `type.h`) instead of `hasFixedSize` and `size`? It seems odd to have to pass these two arguments separately considering they're not really independent (i.e., if one is to specify a specific size, then the intention is presumably to make the mode format fixed size).

stephenchouca commented on include/taco/lower/mode_format_impl.h:
```diff
@@ -172,6 +172,8 @@ class ModeFormatImpl {
   const bool hasInsert;
   const bool hasAppend;
 
+  bool hasFixedSize;
```
Can perhaps replace this and next member with a `Dimension` object? This should also be made const.

stephenchouca commented on include/taco/lower/mode_format_impl.h:
```diff
@@ -66,13 +66,13 @@ class ModeFormatImpl {
   /// Create a copy of the mode type with different properties.
   virtual ModeFormat copy(
-      std::vector<ModeFormat::Property> properties) const = 0;
+      std::vector<ModeFormat::Property> properties, int size = 0) const = 0;
```
`size` can perhaps be of `Dimension` type (see above)?

stephenchouca commented:
Also, the singleton mode format is not restricted to any fixed size; it is meant to be able to store dimensions of any size (consider the column coords array of COO for instance).

fredrikbk commented:
Closing as the ideas have to be integrated with our new blocked types.
